partition-manager/partition-manager/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+#![cfg_attr(not(test), no_std)]
+#![doc = include_str!(concat!("../", env!("CARGO_PKG_README")))]
+
+use core::{fmt::Debug, marker::PhantomData};
+use embassy_sync::{
+    blocking_mutex::raw::{NoopRawMutex, RawMutex},
+    mutex::Mutex,
+};
+
+mod ext;
+
+#[cfg(test)]
+mod test;
+
+/// Manager for the partitions for a storage device.
+///
+/// Manages concurrent device access and ties lifetime to partitions.
+pub struct PartitionManager<F, M: RawMutex = NoopRawMutex> {
+    storage: Mutex<M, F>,
+}
+
+/// Partition of a disk.
+///
+/// If the underlying disk implements [embedded_storage_async::nor_flash::NorFlash] or [block_device_driver::BlockDevice], this partition will too.
+/// Requires the features `esa` and/or `bdd` to be enabled for this crate.
+#[allow(unused)]
+pub struct Partition<'a, F, MARKER, M: RawMutex = NoopRawMutex> {
+    storage: &'a Mutex<M, F>,
+    offset: u32,
+    size: u32,
+    _marker: PhantomData<MARKER>,
+}
+
+impl<F, M: RawMutex> Partition<'_, F, RW, M> {
+    /// Temporarily convert a non-mutable reference to a writable partition into a read-only partition.
+    pub const fn readonly(&self) -> Partition<'_, F, RO, M> {
+        Partition {
+            storage: self.storage,
+            offset: self.offset,
+            size: self.size,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Temporarily convert a writable partition into a read-only partition.
+    pub const fn readonly_mut(&mut self) -> Partition<'_, F, RO, M> {
```

Comment:
I agree with @asasine here, since you do not return an `&mut Parition` but an owned `Partition` you do not need the extra `readonly_mut`. All device accesses are protected through the `Mutex` anyway so only need a non-`mut` reference.

Reply:
A yes, I see what you mean with an owned Partition. Do note that the lifetimes are bound. Then I would prefer only having a &mut self variant to limit the existence of an usable partition to one instance. The &self variant is only used at `esa` `capacity`, but I can live with the code being duplicated there.